# Detecting Pathogen DAN in Sequencing Data 

## System Specification
- **OS:** Ubuntu 22.04.01
- **Virtualization Platform:** Virtualbox 6.5.0
- **Architecture:** x86_64 GNU/Linux
- **Disk Space:**  100 GB
- **Memory (RAM):**  5.3 GB

## Reference Genomes
- Host (Mouse C57BL/6J):
  - https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000001635.27/
- Pathogens:
  1. Helicobacter hepaticus: https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000007905.1/
  2. Staphylococcus aureus: https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000013425.1/
  3. Enterococcus faecalis: https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000393015.1/
  4. Rodentibacter pneumotropicus: https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000730685.1/
  5. Klebsiella oxytoca: https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_003812925.1/
  6. Rodentibacter heylii: https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_010587025.1/

Downloaded .fna files using 'ncbi-datasets' tool:
  - `datasets download genome accession GCF_000007905.1,GCF_000393015.1,GCF_003812925.1,GCF_000013425.1,GCF_000730685.1,GCF_010587025.1 --include genome`
  - Combine pathogen data: `cat GCF_000007905.1_ASM790v1_genomic.fna
GCF_000393015.1_EntefaecT5V1_genomic.fna GCF_003812925.1_ASM381292v1_genomic.fna
GCF_000013425.1_ASM1342v1_genomic.fna GCF_000730685.1ASM73068v1_genomic.fna
GCF_010587025.1_ASM1058702v1_genomic.fna> combined_pathogen_genomes.fna
`
- Excursion FASTA Files:
  - FASTA (.fna) files have the following structure:
    - Lines starting with '>' indicate the beginning of a sequence record (incl unqiue identifier).
    - The actual sequence data follows this indicator line (can be multiple lines)
    ```bash
    >NC_004917.1 Helicobacter hepaticus ATCC 51449, complete sequence
    CATTAAACCAAGTATAAAATCTATAAATTATCTTTA...
    ```
- Indexing:
  - BWA indexing is the process of creating an efficient data structure from a reference genome, allowing the BWA algorithm to rapidly align short DNA sequences during high-throughput sequencing analysis.
   - `bwa index -p pathogen_combined_index combined_pathogen_genomes.fna`
   - `bwa index -p mouse_index GCF_000001635.27_GRCm39_genomic.fna` (~18 hours!)

## Experimental Data

Download DNA sequencing data from experiments, onbtained from feces; ideally containing full genomes (microbial and host components):
  - https://www.ebi.ac.uk/ena/browser/view/SRX3198644:  `fasterq-dump --outdir . --gzip SRX3198644`
    - Only 74.9 kB --> 220 reads (`wc -l SRX3198644.fastq | awk '{print $1/4}'`)

  - https://www.ncbi.nlm.nih.gov/sra/SRX22381836[accn]: `fastq-dump --outdir . --gzip --split-files SRR26681942`
    - Both files 10.0 GB each (!)  --> two files as data was generated by pair-end sequencing
    - -> 22.4M reads each set --> heavy on computation --> create subsets for a general overview

  - Excursion: FASTQ Files:
    - Contains sequence data and quality information, as produced by sequencing machines
    - Each entry conists of four lines:
        - Sequence Indentifier Line: Starts with **'@'** followed by unique identifier for the read, may contain additional information like sequencing instrumet, sample details, read length, etc.
        - Sequence Line: Containts actual nucleotide sequence of DNA/RNA read.
        - Quality Score Identifier Line: Start with **'+'** and usually mirrors the sequence identifier line.
        - Quality Score Line: Contains ASCII-encoded quality scores, representing the confidence/accuracy of each base call.
      ```bash
      @SRX3198644.1 1 length=45
      TTGTTGAACTGGCTCTTTTTCGCAATCCCGCTGTAAGTACTGTCT
      +SRX3198644.1 1 length=45
      AAAAAEEEEEEEEEEEEEEEEEEEAEEAEEEEEEEEEEEEEEEEE
      ```

  -  Generating subsets of ~1M reads each using 'seqtk':
      - `seqtk sample -s 42 SRR26681942_1.fastq 0.05 > subset_SRR26681942_1.fastq`
      - `seqtk sample -s 42 SRR26681942_2.fastq 0.05 > subset_SRR26681942_2.fastq`
      - '42' is a seed value (making random sampling reproducale), new subsets contain 5% of total reads.

## Aligning Seq Data to Genomes

Pipeline:
1. align reads from exp (feces) to mouse genome
2. filter unmapped reads
3. align those reads to pathogens

Simple script reading a single FASTQ: aling_filter_simple.sh

```bash  
#!/bin/bash

exp="SRX3198644"

mouse_index="../1_index/mouse_index"
pathogen_index="../1_index/pathogen_combined_index"
input_reads="../2_exp_fastq/${exp}.fastq"

mkdir -p "$exp"

# Step 1: Align against the mouse genome
echo -e "-----\nStep 1: Aligning against the mouse genome...\n-----"
bwa mem -t 4 "$mouse_index" "$input_reads" > "$exp/$exp_aligned_to_mouse.sam"
echo -e "-----\nStep 1 completed.\n-----"

# Step 2: Filter unmapped reads
echo -e "-----\nStep 2: Filtering unmapped reads...\n-----"
samtools view -b -f 4 "$exp/$exp_aligned_to_mouse.sam" > "$exp/$exp_unmapped_to_mouse.bam"
echo "Step 2 completed.\n"

# Step 3: Convert to FASTQ
echo -e "-----\nStep 3: Conversion to FASTQ...\n-----"
samtools fastq -n -0 "$exp/$exp_unmapped_to_mouse.fastq" "$exp/$exp_unmapped_to_mouse.bam"
echo -e "-----\nStep 3 completed.\n-----"

# Align unmapped reads against the pathogens
echo -e "-----\nStep 4: Align against pathogens...\n-----"
bwa mem -t 4 "$pathogen_index" "$exp/$exp_unmapped_to_mouse.fastq" > "$exp/$exp_aligned_to_pathogen.sam"
echo -e "-----\nStep 4 completed.\n-----"

# Convert SAM to BAM
echo -e "-----\nStep 5: Converting SAM to BAM...\n-----"
samtools view -b -o "$exp/$exp_aligned_to_pathogen.bam" "$exp/$exp_aligned_to_pathogen.sam"
echo -e "-----\nStep 5 completed.\n-----"

# Sort BAM file
echo -e "-----\nStep 6: Sorting BAM file...\n-----"
samtools sort -o "$exp/$exp_aligned_to_pathogen_sorted.bam" "$exp/$exp_aligned_to_pathogen.bam"
echo -e "-----\nStep 6 completed.\n-----"

# Index the sorted BAM file
echo -e "-----\nStep 7: Indexing the sorted BAM file...\n-----"
samtools index "$exp/$exp_aligned_to_pathogen_sorted.bam"
echo -e "-----\nStep 7 completed.\n-----"

# Generate index statistics for the final BAM file
echo -e "-----\nStep 8: Generating index statistics...\n-----"
samtools idxstats "$exp/$exp_aligned_to_pathogen_sorted.bam" > "$exp/$exp_aligned_to_pathogen_idxstats.txt"
echo -e "-----\nStep 8 completed.\n-----"

# Calculate total number of reads (excl. supplementary (flag 2048) and secondary reads (flag 256)).
total_reads=$(samtools view -c -F 2304 "$exp/$exp_aligned_to_mouse.sam")

# Calculate number of reads not mapped to mouse
unmapped_to_mouse_reads=$(samtools view -c -f 4 "$exp/$exp_aligned_to_mouse.sam")

# Calculate number of reads mapped to pathogens (excl. supplementary (flag 2048) and secondary reads (flag 256) along with unmapped (flag 4))
mapped_to_pathogen_reads=$(samtools view -c -F 2308 "$exp/$exp_aligned_to_pathogen_sorted.bam")

# Calculate number of reads not mapped to mouse or pathogens
unmapped_to_either_reads=$((unmapped_to_mouse_reads - mapped_to_pathogen_reads))

# Calculate percentages
percentage_not_mapped_to_mouse=$(awk "BEGIN {printf \"%.2f\", (${unmapped_to_mouse_reads}/${total_reads})*100}")
percentage_mapped_to_pathogen=$(awk "BEGIN {printf \"%.2f\", (${mapped_to_pathogen_reads}/${total_reads})*100}")
percentage_not_mapped_to_either=$(awk "BEGIN {printf \"%.2f\", (${unmapped_to_either_reads}/${total_reads})*100}")

# Display results
echo "Total Reads: $total_reads (100,00%)"
echo "------"
echo "Reads not mapped to mouse: $unmapped_to_mouse_reads (${percentage_not_mapped_to_mouse}%)"
echo "Reads mapped to pathogens: $mapped_to_pathogen_reads (${percentage_mapped_to_pathogen}%)"
echo "Reads not mapped to mouse or pathogen: $unmapped_to_either_reads (${percentage_not_mapped_to_either}%)"

# Redirect results ro text file
echo -e "Total Reads: $total_reads (100,00%)\n\------\nReads not mapped to mouse: $unmapped_to_mouse_reads (${percentage_not_mapped_to_mouse}%)\nReads mapped to pathogens: $mapped_to_pathogen_reads (${percentage_mapped_to_pathogen}%)\nReads not mapped to mouse or pathogen: $unmapped_to_either_reads (${percentage_not_mapped_to_either}%)" > "$exp/$exp_alignment_results.txt"

```


#### BWA (Burrows-Wheeler Aligner):
Purpose:
- BWA is a software package used for aligning short DNA sequences against a large reference genome.
- It employs the Burrows-Wheeler Transform (BWT) algorithm to efficiently align short reads to a reference sequence.

Usage in the Script:
- In the script, BWA is used to align the experimental short DNA sequences (reads) against two different reference genomes: the mouse genome and a combined index for various pathogens.
- The `bwa mem` command is specifically used for DNA sequence alignment using the mem algorithm, suitable for short reads.
  
#### Samtools:
Purpose:
- Samtools is a suite of programs for interacting with high-throughput sequencing data in SAM (Sequence Alignment/Map) and BAM (Binary Alignment/Map) formats.
- It provides various utilities for manipulating, viewing, and analyzing sequence alignment data.
- 
Usage in the Script:
- In the script, Samtools is primarily used for two tasks:
  1. Filtering Unmapped Reads:
    - samtools view is used to filter reads based on specific criteria, such as selecting only unmapped reads ( `-f 4 ` flag).
    - The filtered reads are then redirected to a new BAM file (unmapped_to_mouse.bam).
      
#### Overall Workflow:
1. Alignment to Mouse Genome:
    - BWA is used to align short DNA reads from the experiment to the mouse genome, producing a SAM file (aligned_to_mouse.sam).

2. Filtering Unmapped Reads:
    - Samtools is employed to filter out unmapped reads from the SAM file, resulting in a new BAM file (unmapped_to_mouse.bam).

3. Conversion to BAM Format:
    - Samtools is used to explicitly convert the BAM file containing unmapped reads to BAM format.

4. Alignment to Pathogens:
   - BWA is used again to align the unmapped reads (now in BAM format) to a combined index for various pathogens, generating a BAM file (aligned_to_pathogen.bam).

